caffe2/operators/gather_op.h

```diff
@@ -6,11 +6,133 @@
+    if (idx < 0) {
+      idx = idx + indexing_axis_dim;
+    }
+    CAFFE_ENFORCE(
+        0 <= idx && idx < indexing_axis_dim,
+        "INDICES element is out of DATA bounds, id=",
+        idx,
+        " axis_dim=",
+        indexing_axis_dim);
+  }
+}
+
+// Actual gather implementation - resizes output and copies indexed data.
+template <typename Index, typename Context>
+static bool gather_impl(
+    const Tensor& data,
+    const Tensor& indices,
+    Tensor* output,
+    int axis,
+    Context& context) {
+  // If we endup using it on GPU doing O(N) memcpy is probably not best :)
+  // TODO: implement prefetching if it starts mattering (TF does it)
+
+  size_t item_bytesize = data.meta().itemsize();
+
+  // ONNX allows negative axis to index from the back, valid range: [-r, r].
+  if (axis < 0) {
+    axis = data.dim() + axis;
+  }
+  CAFFE_ENFORCE_GE(data.dim(), axis + 1, "DATA should be at least [axis+1]-D");
+  CAFFE_ENFORCE_GE(axis, 0, "Axis should be non-negative");
+  CAFFE_ENFORCE_LT(axis, data.dim(), "Axis out of range");
+
+  // New shape:
+  //  [data dims before axis] + [indices dims] + [data dims after axis]
+  vector<Index> shape =
+      calc_output_shape_vector<Index>(data.dims(), indices.dims(), axis);
+  output->Resize(shape);
+
+  const Index* idxs = indices.template data<Index>();
+  auto src_base = static_cast<const char*>(data.raw_data());
+  auto out = static_cast<char*>(output->raw_mutable_data(data.meta()));
+
+  auto outer_dims_product = data.size_to_dim(axis);
+  auto block_size = data.size_from_dim(axis + 1);
+  auto block_bytesize = block_size * item_bytesize;
+
+  auto src_indexing_axis_dim = data.dim(axis);
+  auto src_batch_bytesize = data.size_from_dim(axis) * item_bytesize;
+  // Treat indices as a single block even if they have multiple dimensions.
+  // The "gathered batch" is a cumulative result combining indexed blocks.
+  auto N = indices.size();
+  auto gathered_batch_bytesize = N * block_size * item_bytesize;
+
+  check_indexarray_range<Index>(idxs, N, src_indexing_axis_dim);
+
+  // Special-case single-float copy for efficiency
+  if (data.template IsType<float>() && block_size == 1) {
+    for (auto batch = 0; batch < outer_dims_product; ++batch) {
+      const float* src_floats =
+          (const float*)(src_base + batch * src_batch_bytesize);
+      float* dst_floats = (float*)(out + batch * gathered_batch_bytesize);
+
+      for (auto i = 0; i < N; ++i) {
+        auto idx = idxs[i];
+        if (idx < 0) {
+          idx = idx + src_indexing_axis_dim;
+        }
+        dst_floats[i] = src_floats[idx];
+      }
+    }
+  } else {
+    // outer_dims_product specifies how many times we repeat inner dimensions,
+    // so we just iterate over it to cover all outer dimensions.
+    for (auto batch = 0; batch < outer_dims_product; ++batch) {
+      for (auto i = 0; i < N; ++i) {
+        auto idx = idxs[i];
+        if (idx < 0) {
+          idx = idx + src_indexing_axis_dim;
```

Comment:
wait a sec - where does this wrapping come from? we should totally throw on out-of-bounds access. That's what the previous code did.

Note, that wrapping around is for **axis index**, not the index being looked up.

Reply:
The gather_op.h implementation specifically had index wrapping, and not just axis; this is why I adopted it in other places.  However, perhaps that was a mistake? I can remove it everywhere, as i don't see how it makes much sense (but wouldn't want some unexpected failure from changing operator).

Speaking of range checking, should we add it on Cuda side? Because currently CPU version checks bounds with CAFFE_ENFORCE but CUDA kernels do nothing, which probably means out of bounds access.